packages/binding-http/src/http-server.ts

```diff
@@ -205,113 +209,126 @@ export default class HttpServer implements ProtocolServer {
-            let href = base + "/" + this.EVENT_DIR + "/" + eventNamePattern;
+
+          thing.properties[propertyName].forms.push(form);
+          console.debug("[binding-http]",`HttpServer on port ${this.getPort()} assigns '${href}' to Property '${propertyName}'`);
+
+          // if property is observable add an additional form with a observable href
+          if (thing.properties[propertyName].observable) {
+            let href = base + "/" + this.PROPERTY_DIR + "/" + encodeURIComponent(propertyName) + "/" + this.OBSERVABLE_DIR;
             let form = new TD.Form(href, type);
-            ProtocolHelpers.updateEventFormWithTemplate(form, tdTemplate, eventName);
+            form.op = ["observeproperty", "unobserveproperty"];
             form.subprotocol = "longpoll";
-            form.op = ["subscribeevent", "unsubscribeevent"];
-            thing.events[eventName].forms.push(form);
-            console.debug("[binding-http]",`HttpServer on port ${this.getPort()} assigns '${href}' to Event '${eventName}'`);
+            thing.properties[propertyName].forms.push(form);
+            console.debug("[binding-http]",`HttpServer on port ${this.getPort()} assigns '${href}' to observable Property '${propertyName}'`);
           }
-        } // media types
-      } // addresses
+        }
 
-      if (this.scheme === "https") {
-        this.fillSecurityScheme(thing)
-      }
+        for (let actionName in thing.actions) {
+          let actionNamePattern = this.updateInteractionNameWithUriVariablePattern(actionName, thing.actions[actionName].uriVariables);
+          let href = base + "/" + this.ACTION_DIR + "/" + actionNamePattern;
+          let form = new TD.Form(href, type);
+          ProtocolHelpers.updateActionFormWithTemplate(form, tdTemplate, actionName);
+          form.op = ["invokeaction"];
+          let hform : HttpForm = form;
+          if(hform["htv:methodName"] === undefined) {
+            hform["htv:methodName"] = "POST";
+          }
+          thing.actions[actionName].forms.push(form);
+          console.debug("[binding-http]",`HttpServer on port ${this.getPort()} assigns '${href}' to Action '${actionName}'`);
+        }
 
-    } // running
+        for (let eventName in thing.events) {
+          let eventNamePattern = this.updateInteractionNameWithUriVariablePattern(eventName, thing.events[eventName].uriVariables);
+          let href = base + "/" + this.EVENT_DIR + "/" + eventNamePattern;
+          let form = new TD.Form(href, type);
+          ProtocolHelpers.updateEventFormWithTemplate(form, tdTemplate, eventName);
+          form.subprotocol = "longpoll";
+          form.op = ["subscribeevent", "unsubscribeevent"];
+          thing.events[eventName].forms.push(form);
+          console.debug("[binding-http]",`HttpServer on port ${this.getPort()} assigns '${href}' to Event '${eventName}'`);
+        }
+      }
+    }
```

Comment:
I found it difficult to understand the changes from 242 till here. Could you briefly explain what/why did you change here? Thanks!

Reply:
I created a new function called addEndpoint, refactoring the functionality that takes the TD template and turns it into links/href.

It used to be inline to the for loop enumerating addresses. The baseUri if/else logic would have been more complicated, since the add endpoint logic was buried in the for loop enumerating the physical addresses, (With baseUri, we don't care about the local addresses in the TD links, and dont want a client to be distracted/confused by interfaces it wont be able to connect to)

Ye Old Way:
https://github.com/eclipse/thingweb.node-wot/blob/45d8fa739cb6e796c2ee38bf156a977c4f0739bf/packages/binding-http/src/http-server.ts#L188-L278

By having it separated, and public, a developer who wants to do some other crazy thing, can just add their own endpoints.